### PR TITLE
Add support for private incidents

### PR DESCRIPTION
**Pull request by cimi**
Allows marking incidents to signal they contain private information.
This can be done on creation or later when editing the incident.
Once the flag is set to true it cannot be undone.

Private incidents are not omitted when listing all incidents.
The incident summary, impact, severity etc. can still be edited from
 Slack after an incident is made private.

Incident details can no longer be retrieved from the API once an
incident is made private (always returns HTTP 403).

Marking an incident as private does not automatically change the comms
channel privacy settings, this needs to be done manually.

Screenshot 2019-11-04 at 11 52 09

### CODE REVIEW

File: response/slack/action_handlers.py [hunk 1]
```diff
@@ -53,41 +53,53 @@ def handle_create_comms_channel(ac: ActionContext):
...
+            label="Severity",
+            name="severity",
+            optional=True,
+        ),
+    ]
+
+    if not ac.incident.private:
+        dialog_elements.append(
+            SelectWithOptions(
+                [("Yes (CANNOT BE UNDONE)", "yes")],
```
Comment (mattrco) [inline response/slack/action_handlers.py]: Very minor tweak but I'd put `Yes (can't be undone!)` for slightly friendlier text.

File: response/core/serializers.py [hunk 2]
```diff
@@ -124,5 +125,8 @@ def update(self, instance, validated_data):
         instance.summary = validated_data.get("summary", instance.summary)
         instance.severity = validated_data.get("severity", instance.severity)
 
+        if not instance.private:
```
Comment (mattrco) [inline response/core/serializers.py]: This makes sense, I'd probably add a small comment on what this does (only allows setting private to true).

**Comment by mattrco** [comment]
🐛 There's a stray `__init__.pyc` here.

File: tests/test.py [hunk 3]
```diff
@@ -220,3 +221,70 @@ def test_edit_incident(post_from_slack_api, mock_slack):
...
+        report_only=False,
+        private=True,
+        summary="Testing editing incidents - before",
+        impact="Lots",
+        lead=user,
+        severity="1",
+    )
+
+    newsummary = "Testing editing incidents - after"
+    data = {
```
Comment (mattrco) [inline tests/test.py]: May be better to explicitly set private=False in this payload as that's being tested implicitly.

File: response/slack/views.py [hunk 4]
```diff
@@ -69,6 +69,16 @@ def slash_command(request):
             )
         )
 
+    dialog.add_element(
+        SelectWithOptions(
+            [("Yes (CANNOT BE UNDONE)", "yes")],
```
Comment (mattrco) [inline response/slack/views.py]: Same minor comment re. wording

File: tests/test.py [hunk 5]
```diff
@@ -220,3 +221,70 @@ def test_edit_incident(post_from_slack_api, mock_slack):
...
+        "payload": json.dumps(
```
Comment (mattrco) [inline tests/test.py]: Is it worth explicitly trying to set private=False in this payload?

**Comment by mattrco** [comment]
@cimi this is great, thank you!

Although we've locked down the `incident_doc` view, private incidents will still be visible through the API (which is also what our internal UI uses to render data). We'll need to add a filter (`private=False`) to some of the view functions in `core/views.py` to achieve this. Happy to help with this if you'd like 👍

**Comment by cimi** [comment]
Thanks for the review @mattrco! I've addressed your comments 😄

> We'll need to add a filter (private=False) to some of the view functions in `core/views.py` to achieve this. Happy to help with this if you'd like 👍

Should we do this in a separate PR? I'm at a conference today and tomorrow, if you have time to do the API update it would be great - we can release this feature sooner. No worries if not, I'll do it early next week.

**Comment by mattrco** [comment]
@cimi I'm happy to merge this as-is and we can tackle the API later - although it's not quite complete, I think we need to do some test refactoring at the same time as it isn't very clear how the API views are tested vs. the HTML response.

**Pull request merged**